src/main/java/org/testng/internal/TestResult.java

```diff
@@ -320,4 +336,89 @@ public boolean wasRetried() {
   public void setWasRetried(boolean wasRetried) {
     this.m_wasRetried = wasRetried;
   }
+
+  public List<ITestNGMethod> getSkipCausedBy() {
+    if (this.m_status != SKIP || skipAnalysed) {
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+    skipAnalysed = true;
+    //check if there were any config failures
+    Set<ITestResult> skippedConfigs = m_context.getFailedConfigurations().getAllResults();
+    for (ITestResult skippedConfig : skippedConfigs) {
+      if (isGlobalFailure(skippedConfig) || isRelated(skippedConfig)) {
+        //If there's a failure in @BeforeTest/@BeforeSuite/@BeforeClass
+        //then the reason is most often just one method.
+        skippedDueTo.add(skippedConfig.getMethod());
+        return Collections.unmodifiableList(skippedDueTo);
+      }
+      if (belongToSameGroup(skippedConfig)) {
+        //If its @BeforeGroups then there's a chance that there could be more than one
+        //method. So lets add everything.
+        skippedDueTo.add(skippedConfig.getMethod());
+      }
+    }
+    if (!skippedDueTo.isEmpty()) {
+      //If we found atleast one skipped due to reason, then its time to return back.
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+    //Looks like we didnt have any configuration failures. So some upstream method perhaps failed.
+    List<String> upstreamMethods = Arrays.asList(m_method.getMethodsDependedUpon());

```

Comment:
For perf reason, maybe we can postpone the array to list transformation (here and on the line 368)

Reply:
Fixed.